samples/solutions/stingy_connection.py

```diff
@@ -0,0 +1,275 @@
+logger = logging.getLogger(__name__)
+logger.addHandler(sample_log_handler)
+
+
+def get_type_name(e):
+    return type(e).__name__
+
+
+def process_memory():
+    process = psutil.Process(os.getpid())
+    mem_info = process.memory_info()
+    return mem_info.rss
+
+
+# decorator function
+def profile(func):
+    def wrapper(*args, **kwargs):
+        mem_before = process_memory()
+        result = func(*args, **kwargs)
+        mem_after = process_memory()
+        print(
+            "{}:consumed memory: {:,}, {}, {}".format(
+                func.__name__, mem_before, mem_after, mem_after - mem_before
+            )
+        )
+
+        return result
+
+    return wrapper
+
+
+class Application(object):
+    async def initiate(self):
+
+        self.transient_errors = [
+            exceptions.OperationCancelled,
+            exceptions.OperationTimeout,
+            exceptions.ServiceError,
+            exceptions.ConnectionFailedError,
+            exceptions.ConnectionDroppedError,
+            exceptions.NoConnectionError,
+            exceptions.ClientError,  # TODO Only TLSEsxhangeError is actually worth retrying
+        ]
+
+        self.message_queue = asyncio.Queue()
+        self.message_list = {}
+        self.task_list = []
+        self.device_client = None
+
+    async def create_client(self, conn_str):
+        try:
+            # Create a Device Client
+            self.device_client = IoTHubDeviceClient.create_from_connection_string(conn_str)
+            # Attach the connection state handler
+            self.device_client.on_connection_state_change = self.handle_on_connection_state_change
+        except Exception as e:
+            self.log_error_and_print(
+                "Caught exception while trying to attach handler : {}".format(get_type_name(e))
+            )
+
+    async def handle_on_connection_state_change(self):
+        self.log_info_and_print(
+            "handle_on_connection_state_change fired. Connected status : {}".format(
+                self.device_client.connected
+            )
+        )
+        if self.device_client.connected:
+            self.log_info_and_print("Connected connected_event is set...")
+        else:
+            self.log_info_and_print("Disconnected connected_event is set...")
+
+    def is_retryable(self, exc):
+        if type(exc) in self.transient_errors:
+            return True
+        else:
+            return False
+
+    def log_error_and_print(self, s):
+        logger.error(s)
+        print(s)
+
+    def log_info_and_print(self, s):
+        logger.info(s)
+        print(s)
+
+    async def enqueue_message(self):
+        message_id = 0
+        while True:
+            msg = Message("current wind speed ")
+            msg.message_id = message_id
+            msg.content_type = "application/json"
+            self.log_info_and_print("Created a message...")
+            self.message_queue.put_nowait(msg)
+            self.message_list[message_id] = msg
```

Comment:
>elf.message_list[message_id] = msg [](http://example.com/codeflow?start=13&length=34)

this list grows and never empties. What is it for?

Reply:
left over code